### PR TITLE
Use vswhere to search for vs2017+ install directory to locate msbuild

### PR DESCRIPTION
**Pull request by yuyoyuppe**
FAKE assumes VS is installed in ProgramFilesX86, which wasn't the case for me. We could use `vswhere` utility with reliable location to determine VS2017+ installation path.

### CODE REVIEW

**Comment by realvictorprm** [comment]
Good job @yuyoyuppe! 😃

File: src/app/FakeLib/MSBuildHelper.fs [hunk 1]
```diff
@@ -96,6 +96,15 @@ let msBuildExe =
             ]
             defaultArg (sources |> List.choose id |> List.tryHead) "xbuild"
         | false, _ ->
+            let getVSPathFromVSWhere ver =
+                ExecProcessAndReturnMessages(fun info ->
```
Comment (matthid) [inline src/app/FakeLib/MSBuildHelper.fs]: I think this throws when the `VSWhere` file doesn't exist. Is this available on all VS Versions? What if no Visual Studio is installed?

**Comment by matthid** [comment]
Good change, only one thing: Can you check what happens when no Visual Studio is installed or VSWhere is missing for some reason? Or investigate if VSWhere is always available.

Thank you.

**Comment by matthid** [comment]
Additionally we need to update the new FAKE 5 modules as well (not only the Obsolete ones). They can be found in `Fake.DotNet.MsBuild` and `Fake.Core.Environment`.

**Comment by wallymathieu** [comment]
You could also use the [nuget package for vswhere](https://www.nuget.org/packages/vswhere) so that you know that it exists regardless (but you would then need to take a dependency on that package).

**Comment by matthid** [comment]
I'd be fine with that dependency, considering that it will be local to the 'Fake.DotNet.MsBuild' module. we probably would disable that feature in fakelib in that scenario

**Comment by wallymathieu** [comment]
In another project we have done:
`build.cmd`:
```cmd
...
for /f "usebackq tokens=*" %%i in (`%~dp0\packages\vswhere\tools\vswhere -latest -products * -requires Microsoft.Component.MSBuild -property installationPath`) do (
  set VSInstallDir=%%i
)

IF NOT EXIST "%VSInstallDir%\MSBuild\15.0\Bin\MSBuild.exe" (
  exit /b -1
)

packages\build\FAKE\tools\FAKE.exe build.fsx %* "VSInstallDir=%VSInstallDir%"
```

and then in build.fsx
```fsharp
let visualStudioInstallDir = getBuildParamOrDefault "VSInstallDir" @"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\"
...
#if MONO
let extraCscParams = id
#else
let extraCscParams (parameters:CscHelper.CscParams) = 
  { parameters with ToolPath = visualStudioInstallDir </> @"MSBuild\15.0\Bin\Roslyn\csc.exe" }
#endif
```
in order to find `csc.exe` (from https://github.com/fsprojects/SyntacticVersioning/pull/24)

**Comment by yuyoyuppe** [comment]
@matthid ah, sorry, completely forgot! :( I'll try fixing the mentioned issues this weekend.

**Comment by matthid** [comment]
Is this ready (it is red)?

**Comment by yuyoyuppe** [comment]
@matthid I think there's a cleaner way to achieve the same result: just run `fake` from VS' native command tool prompt and its location will be detected, so imo this patch is unnecessary.

**Comment by wallymathieu** [comment]
VS native command tool prompt is not available on all platforms. It's only available when you have installed VS (something that you might not want to do on a build server).

**Comment by matthid** [comment]
I think this would be viable as long as it still works when `vswhere.exe` is not found. Because as @wallymathieu said most people don't use the VS command line. Personally, I use the VS-Code bash terminal for example.

But if there is no interest in finishing this. Should we close the PR?

**Comment by matthid** [comment]
Looks like it has been superseded by https://github.com/fsharp/FAKE/pull/2077

**Pull request closed without merging**